### PR TITLE
Do not prefetch when possible

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Only prefetch unless we require it. That is when:
- Iterable datasets are used, where we need to know when it will be exhausted to run validation.
- Interbatch parallelism is used, where no prefetching would be silly.

This is good for FFCV because they already do prefetching internally.

We no longer error on 0-length map-style datasets as this is inconsistent with iterable datasets.

Part of #11538
Follow-up to #11606
Reverts #1280

### Does your PR introduce any breaking changes? If yes, please list them.

None knowingly

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @ninginthecloud @akihironitta

### CODE REVIEW

File: pytorch_lightning/utilities/fetching.py [hunk 1]
```diff
@@ -172,8 +175,6 @@ def _attach_data_fetcher_fn(loader: DataLoader) -> None:
 
     def __iter__(self) -> "AbstractDataFetcher":
         self.reset()
-        self._attach_data_fetcher()
-        _patch_dataloader_get_iterators()
```
Comment (carmocca) [inline pytorch_lightning/utilities/fetching.py:176]: I believe these only need to be called once on setup, so moved them

File: tests/loops/test_loops.py [hunk 2]
```diff
@@ -648,16 +648,12 @@ def train_dataloader(self):
...
-                # TODO: the following "-1" offset will be fixed by
-                #   https://github.com/PyTorchLightning/pytorch-lightning/pull/8578
                 "completed": n_epochs - 1,
             },
             "current": {
                 "ready": n_epochs,
                 "started": n_epochs,
                 "processed": n_epochs,
-                # TODO: the following "-1" offset will be fixed by
-                #   https://github.com/PyTorchLightning/pytorch-lightning/pull/8578
```
Comment (carmocca) [inline tests/loops/test_loops.py:660]: Unrelated TODO which does not apply and was forgotten in #8578

File: tests/loops/test_loops.py [hunk 3]
```diff
@@ -956,8 +952,6 @@ def val_dataloader(self):
     # totals are increased by 1 (the failed batch which never completed)
     expected = state_dict.copy()
 
-    # TODO: `is_last_batch` is not correct on reload, the next line should not be necessary
```
Comment (carmocca) [inline tests/loops/test_loops.py:959]: I remember being puzzled about why this did not work back when I wrote this.

Turns out the problem was that we were not restoring the `fetched` value on restart

File: tests/utilities/test_fetching.py [hunk 4]
```diff
@@ -30,55 +29,66 @@
...
+        yield 2
+        yield 3
 
-    class IterDataset(IterableDataset):
-        def __iter__(self):
-            yield 1
-            yield 2
-            yield 3
-
-    for prefetch_batches in range(5):
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:43]: This loop is now a parametrization so the indentation changed.

File: tests/utilities/test_fetching.py [hunk 5]
```diff
@@ -30,55 +29,66 @@
...
+class SizedDataset(Dataset):
+    def __len__(self):
+        return 3
+
+    def __getitem__(self, idx):
+        return idx + 1
+
+
+@pytest.mark.parametrize("use_combined_loader", [False, True])
+@pytest.mark.parametrize("dataset_cls", [IterDataset, SizedDataset])
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:48]: Now testing sized datasets too

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -81,6 +81,13 @@ def dataloaders(self) -> Sequence[DataLoader]:
             raise RuntimeError("Dataloaders should be available.")
         return dataloaders
 
+    @property
+    def prefetch_batches(self) -> int:
+        batches = self.trainer.num_test_batches if self.trainer.testing else self.trainer.num_val_batches
+        iterable_dataset = batches[self.current_dataloader_idx] == float("inf")
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:87]: An iterable dataset can have a length.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 7]
```diff
@@ -81,6 +81,13 @@ def dataloaders(self) -> Sequence[DataLoader]:
...
+        inter_batch_parallelism = os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1"
+        return 1 if iterable_dataset or inter_batch_parallelism else 0
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:89]:
```suggestion
        un_sized_dataset = batches[self.current_dataloader_idx] == float("inf")
        inter_batch_parallelism = os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1"
        return 1 if un_sized_dataset or inter_batch_parallelism else 0
```

File: pytorch_lightning/loops/fit_loop.py [hunk 8]
```diff
@@ -149,6 +149,12 @@ def restarting(self, restarting: bool) -> None:
         restarting &= finished_before_on_train_end
         Loop.restarting.fset(self, restarting)  # call the parent setter
 
+    @property
+    def prefetch_batches(self) -> int:
+        iterable_dataset = self.trainer.num_training_batches == float("inf")
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:154]: Same as above.

File: pytorch_lightning/utilities/fetching.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:176]: If the iterators are being re-created, we won't be patching the correct new ones.

Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:176]: But you might be right. Worth to check.

**Review by tchaton — APPROVED** [review]
LGTM ! Some comments.

File: pytorch_lightning/utilities/fetching.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:176]: `_patch_dataloader_get_iterators` is a simple monkeypatch on the class unrelated to the instances

and `_attach_data_fetcher` only uses the `self.dataloader` property which is read-only and only gets set during `setup` (the place where we call these 2 now)

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 9]
```diff
@@ -81,6 +81,13 @@ def dataloaders(self) -> Sequence[DataLoader]:
             raise RuntimeError("Dataloaders should be available.")
         return dataloaders
 
+    @property
+    def prefetch_batches(self) -> int:
+        batches = self.trainer.num_test_batches if self.trainer.testing else self.trainer.num_val_batches
+        is_unsized = batches[self.current_dataloader_idx] == float("inf")
+        inter_batch_parallelism = os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1"
```
Comment (kaushikb11) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:88]: We should probably mention it in the Data section of the docs.

Comment (justusschock) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:88]: is there no other way to trigger it? I don't like features only to be accessible by environment variables

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/data.py [hunk 10]
```diff
@@ -89,17 +89,13 @@ def has_iterable_dataset(dataloader: DataLoader) -> bool:
...
-            If the length of Dataloader is 0, as it requires at least one batch
-    """
-
+    infinite dataloader."""
     try:
         # try getting the length
         if len(dataloader) == 0:
-            raise ValueError("`Dataloader` returned 0 length. Please make sure that it returns at least 1 batch")
+            rank_zero_warn(
+                f"`{dataloader.__class__.__name__}` returned 0 length. Please make sure this was your intention."
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/data.py:97]: I think `dataloader.__class__.__name__` will return `DataLoader` only. You mean `dataloader.dataset` here?

File: pytorch_lightning/utilities/data.py [hunk 11]
```diff
@@ -122,30 +118,27 @@ def has_len_all_ranks(
...
-                    "Total length of `Dataloader` across ranks is zero, but local rank has zero length."
-                    " Please be cautious of uneven batch length."
+                    f"Total length of `{dataloader.__class__.__name__}` across ranks is zero, but local rank has zero"
+                    " length. Please be cautious of uneven batch length."
                 )
                 has_len = False
             else:
                 raise MisconfigurationException(
-                    "`Dataloader` within local rank has zero length. Please make sure that it returns at least 1 batch."
+                    f"`{dataloader.__class__.__name__}` within local rank has zero length."
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/data.py:140]: same here.

File: docs/source/guides/data.rst [hunk 12]
```diff
@@ -393,6 +393,9 @@ option when using sequential data.
     to ``limit_{mode}_batches``, if it is set to 1.0 it will run for the whole dataset, otherwise it will throw an exception.
     Here ``mode`` can be train/val/test/predict.
 
+When iterable datasets are used, Lightning will pre-fetch 1 batch (in addition to the current batch) so it can detect
+when the training will stop and run validation if necessary.
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/guides/data.rst:397]: It can happen at any point in training:

https://github.com/PyTorchLightning/pytorch-lightning/blob/02ccd874b9eda885df7752710e605e259d579d9c/pytorch_lightning/loops/epoch/training_epoch_loop.py#L488-L489

File: pytorch_lightning/utilities/data.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/utilities/data.py:97]: This maintains the previous message and follows what's being checked in the code: `len(dataloader)==0`

File: docs/source/guides/data.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/guides/data.rst:397]: then maybe `when the validation loop needs to be run.`

File: pytorch_lightning/utilities/data.py (hunk 10 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:97]: but isn't that upto the dataset to decide?

if not maybe just update to `DataLoader` since `dataloader.__class__.__name__` is static.

Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:97]: oh I see.. thanks :)

File: docs/source/guides/data.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/guides/data.rst:397]: while the comment here is not wrong, the real reason we have the prefetching is actually to avoid starting a new epoch that doesn't have any batches left in the dataloader. Relying on a StopIteration check is not possible for this reason.
